### PR TITLE
Fix the un-hashability of various annotation types, for use in caching generic containers

### PR DESCRIPTION
**Pull request by dmontagu**
skip change file check

Selected Reviewer: @hramezani

### CODE REVIEW

File: tests/test_types.py [hunk 1]
```diff
@@ -5107,6 +5112,16 @@ class A(BaseModel):
...
 
 
+def test_transform_schema():
+    ValidateStrAsInt = Annotated[str, TransformSchema(lambda _s: core_schema.int_schema())]
+
+    class Model(BaseModel):
+        x: Optional[ValidateStrAsInt]
+
+    assert Model(x=None).x is None
+    assert Model(x='1').x == 1
```
Comment (ybressler) [inline tests/test_types.py:5122]: Transformation from a string repr of an int/float/decimal is no longer automatically executed in V2, right? I'm assuming you are overriding this with `TransformSchema(lambda _s: core_schema.int_schema())`?

Also, this test could be rewritten as parameterized if you wanted to go overboard. Perhaps some other edge cases like negative numbers, or floats? (Though would a float get transformed to an int?)

**Review by ybressler — COMMENTED** [review]
> _**Note:** I'm trying to get in the habit of reviewing more code here. Apologies if my comments are n00b._

Code looks straightforward overall. I appreciate the addition of tests. Your PR description mentions inclusion of methods in caching, though I don't see any tests which relate to the caching of these methods. (Unless they are implicit?)

**Comment by dmontagu** [comment]
please review

**Review by hramezani — APPROVED** [review]
